File: BE/HDDT_BE/Models/Professional/ThongTinHangHoaModel.cs
using HDDT_BE.Constants;
using HDDT_BE.Models.Core;
using MongoDB.Bson.Serialization.Attributes;

namespace HDDT_BE.Models.Professional;


// Thong tin hang hoa
public class ThongTinHangHoaModel : Audit, TEntity<string>
{
    public string ma { get; set; }

    public LoaiDichVuYTeModelShort LoaiDichVuYTe{
        get;
        set;
    }
    
    public ThuocModelShort Thuoc{
        get;
        set;
    }
    
    public CommonModelShort DonViTinh{
        get;
        set;
    }
    public string TenHangHoa { get; set; }
    
    
    

    public double DonGia { get; set; } = 0; 
    
    [BsonIgnore]
    public string DonGiaShow 
    {
        get
        {
            return DonGia != null ? FormatMoney.ConvertToMoney(DonGia) : ""; 
        }
    }


    public double ChenhLech
    {
        get
        {
            if (KhongTru)
            {
                return DonGia - GiaMuaVao - PhiTiem - PhiTuVan;
            }
            else
            {
                return DonGia - GiaMuaVao - PhiTiem;
            }
        }
        set{}
    }
    
    
    
    [BsonIgnore]
    public string ChenhLechShow 
    {
        get
        {
            return ChenhLech != null ? FormatMoney.ConvertToMoney(ChenhLech) : ""; 
        }
    }



    
    
    
    public double GiaMuaVao { get; set; }
    
    [BsonIgnore]
    public string GiaMuaVaoShow 
    {
        get
        {
            return GiaMuaVao != null ? FormatMoney.ConvertToMoney(GiaMuaVao) : ""; 
        }
    }
    
    public double GiaLamTron { get; set; }
    
    
    [BsonIgnore]
    public string GiaLamTronShow 
    {
        get
        {
            return GiaLamTron != null ? FormatMoney.ConvertToMoney(GiaLamTron) : ""; 
        }
    }
    
    public double PhiTuVan { get; set; }
    
    [BsonIgnore]
    public string PhiTuVanShow 
    {
        get
        {
            return PhiTuVan != null ? FormatMoney.ConvertToMoney(PhiTuVan) : ""; 
        }
    }
    
    public double PhiTiem { get; set; }
    
    
    [BsonIgnore]
    public string PhiTiemShow 
    {
        get
        {
            return PhiTiem != null ? FormatMoney.ConvertToMoney(PhiTiem) : ""; 
        }
    }
    
    public double SoLuong { get; set; } = 0;

    
    public double TongTien
    {
        get
        {
            return  SoLuong * DonGia;
        }
        set
        {
        }
    }
    
    [BsonIgnore]
    public string TongTienShow 
    {
        get
        {
            return TongTien != null ? FormatMoney.ConvertToMoney(TongTien) : ""; 
        }
    }

    
    public string HoaDonId { get; set; }
    
    
    public string FullName { get; set; }
    
    
    public string UnsignedFullName
    {
        get { return FullName != null ? FormatterString.ConvertToTenKhongDau(FullName) : ""; }
        set { }
    }

    
    public string InvoiceNo { get; set; }
    
    
    public string StatusCode { get; set; }
    
    
    public string TenNguoiMua { get; set; }

    
    public string StatusName { get; set; }
    
    public int? CreatedAtString
    {
        get { return CreatedAt.HasValue ? FormatDate.ConvertDatetimeQG(CreatedAt) : null; }
        set {}
    }
    
    public bool KhongTru { get; set; } = false;

}

public class ThongTinHangHoaShortModel
{
    public string ma { get; set; }

    public LoaiDichVuYTeModelShort LoaiDichVuYTe{
        get;
        set;
    }
    
    public ThuocModelShort Thuoc{
        get;
        set;
    }
    
    public CommonModelShort DonViTinh{
        get;
        set;
    }
    public string TenHangHoa { get; set; }

    public double DonGia { get; set; } 


    public double SoLuong { get; set; } 


    public double TongTien { get; set; } 

    
    public string HoaDonId { get; set; }
    
    
    
    public string UserName { get; set; }
    
    


}

File: FE/src/models/thongTinHangHoaModel.js
import {formatNumber} from "@/models/formatNumberModel";

const sendJson = (item) => {
  //  console.log("LOG SEND JSON : ",  formatNumber.FormatNumber(item.donGia),)
    return {
        ma: item.ma,
        loaiDichVuYTe : item.loaiDichVuYTe,
        thuoc : item.thuoc,
        donViTinh : item.donViTinh,
        tenHangHoa : item.tenHangHoa,
        soLuong : formatNumber.FormatNumber(item.soLuong),
        donGia :  formatNumber.FormatNumber(item.donGia),
        phiTuVan :  formatNumber.FormatNumber(item.phiTuVan),
        phiTiem : formatNumber.FormatNumber(item.phiTiem),
        khongTru : item.khongTru
    }
}

const sendJsonFirst = (item,  ma ,soLuong , donGia, khongTru) => {
    console.log("LOG SENDJSON FIRSE : ", khongTru)
    return {
        ma: ma,
        loaiDichVuYTe : item.loaiDichVuYTe,
        thuoc : item.thuoc,
        donViTinh : item.donViTinh,
        tenHangHoa : item.tenHangHoa,
        soLuong : formatNumber.FormatNumber(soLuong),
        donGia :  formatNumber.FormatNumber(donGia),
        phiTuVan :  formatNumber.FormatNumber(item.phiTuVan),
        phiTiem : formatNumber.FormatNumber(item.phiTiem),
        khongTru : khongTru
    }
}

const sendJsonLeftSection = (item) => {
    console.log("LOG DON GIA : ", item.donGia)
    return {
        ma: item.ma,
        loaiDichVuYTe : item.loaiDichVuYTe,
        thuoc : item.thuoc,
        donViTinh : item.donViTinh,
        tenHangHoa : item.tenHangHoa,
        soLuong : formatNumber.FormatNumber(item.soLuong) - 1,
        donGia :  formatNumber.FormatNumber(item.donGia),
        phiTuVan :  formatNumber.FormatNumber(item.phiTuVan),
        phiTiem : formatNumber.FormatNumber(item.phiTiem),
        khongTru : item.khongTru
    }
}




const baseJson = () => {
    return {
        ma: null,
        thuoc : null ,
        loaiDichVuYTe : null,
        donViTinh : null ,
        tenHangHoa : null,
        soLuong : 1,
        donGia : null,
        tongTien : null,
        phiTuVan : null,
        phiTiem : null,
        khongTru : true,
    }
}

export const thongTinHangHoaModel = {
    baseJson , sendJson,sendJsonFirst ,sendJsonLeftSection
}
